src/probability.ts

```diff
@@ -0,0 +1,186 @@
+            [0,0,0,0,0,0,0,0,0,0],
+            [0,0,0,0,0,0,0,0,0,0],
+            [0,0,0,0,0,0,0,0,0,0],
+            [0,0,0,0,0,0,0,0,0,0],
+        ]
+        for (let len of ships){
+            let gamestate:GameState = new GameState({});
+            gamestate.board = board;
+            gamestate.eliminateSunkenShips();
+            gamestate.eliminateNeighboursOfSunken();
+            gamestate.eliminateSunkenShips();
+            gamestate.eliminateNeighboursOfSunken();
+            // We have to do the above twice because we might be able to sink additional ships 
+            // after eliminating neighbours of already sunken ships
+
+            let found:boolean = false
+            let count:number = 0;
+            while (!found){
+                if (count>1000) {
+                    console.log("while loop exceeded 1000 iterations, so default shipplacement was returned");
+                    return [
+                        { StartingSquare: { Row: "A", Column: 1 }, EndingSquare : { Row: "A", Column: 5 } },
+                        { StartingSquare: { Row: "C", Column: 1 }, EndingSquare : { Row: "C", Column: 4 } },
+                        { StartingSquare: { Row: "E", Column: 1 }, EndingSquare : { Row: "E", Column: 3 } },
+                        { StartingSquare: { Row: "G", Column: 1 }, EndingSquare : { Row: "G", Column: 3 } },
+                        { StartingSquare: { Row: "I", Column: 1 }, EndingSquare : { Row: "I", Column: 2 } },
+                    ];
+                }
+                count++;
+                let pos:Position = gamestate.randomDraw(true);
+                let rowDirections:number[] = [0,0,1,-1];
+                let columnDirections:number[] = [1,-1,0,0];
+                let directions:number[] = Probability.shuffleArray([0,1,2,3]);
+                for (let i:number = 0; i<4; i++){
+                    let dr:number = rowDirections[directions[i]];
+                    let dc:number = columnDirections[directions[i]];
+
+                    let valid:boolean = true;
+                    for (let i:number = 0; i<len; i++){
+                        if (!gamestate.isValidTarget(new Position({"Row":GameState.numberToLetter[pos.row+i*dr], "Column":pos.column+1+i*dc}))){
+                            valid = false;
+                        }
+                    }
+                    if (valid) {
+                        found = true;
+                        answer.push({StartingSquare:{Row:GameState.numberToLetter[pos.row], Column:pos.column+1},
+                                    EndingSquare:{Row:GameState.numberToLetter[pos.row+(len-1)*dr], Column:pos.column+1+(len-1)*dc}
+                                });
+                        for (let j:number = 0; j<len; j++){
+                            board[pos.row+j*dr][pos.column+j*dc] = 2;
+                        }
+                    }
+                }
+            }
+        }
+        return answer;
+    }
```

Comment:
Like a few of the methods here, this is _massive_.

If there was a bug in this specific bit, how long do you reckon it would take you to find it? How likely would it be that you could accidentally break it and not notice?

This really should be split into a number of smaller methods.